src/common/lib/python/common_extension.cc

```diff
@@ -323,20 +322,40 @@ static int PyTask_init(PyTask *self, PyObject *args, PyObject *kwds) {
       Py_DECREF(data);
     }
   }
-  /* Set the resource vector of the task. */
-  if (resource_vector != NULL) {
-    CHECK(PyList_Size(resource_vector) == ResourceIndex_MAX);
-    for (int i = 0; i < ResourceIndex_MAX; ++i) {
-      PyObject *resource_entry = PyList_GetItem(resource_vector, i);
-      TaskSpec_set_required_resource(g_task_builder, i,
-                                     PyFloat_AsDouble(resource_entry));
+  /* Set the resource requirements for the task. */
+  bool found_CPU_requirements = false;
+  PyObject *key, *value;
+  Py_ssize_t position = 0;
+  if (resource_map != NULL) {
+    if (!PyDict_Check(resource_map)) {
+      PyErr_SetString(PyExc_TypeError, "resource_map must be a dictionary");
+      return -1;
     }
-  } else {
-    for (int i = 0; i < ResourceIndex_MAX; ++i) {
-      TaskSpec_set_required_resource(g_task_builder, i,
-                                     i == ResourceIndex_CPU ? 1.0 : 0.0);
+    while (PyDict_Next(resource_map, &position, &key, &value)) {
+      if (!PyUnicode_Check(key)) {
+        PyErr_SetString(PyExc_TypeError,
+                        "the keys in resource_map must be strings");
+        return -1;
+      }
+      if (!(PyFloat_Check(value) || PyLong_Check(value))) {
+        PyErr_SetString(PyExc_TypeError,
+                        "the values in resource_map must be floats");
+        return -1;
+      }
+
+      std::string resource_name(PyBytes_AsString(PyUnicode_AsASCIIString(key)),
```

Comment:
Careful here! PyUnicode_AsASCIIString creates a new reference and needs to be decrefed.

Reply:
Good catch!